<launch>

	<arg name="epuck2_address"/>
	<arg name="epuck2_name" default="epuck2_robot_0"/>
	<arg name="cam_en" default="false"/>
	<arg name="imu_en" default="true"/>
	<arg name="prox_en" default="true"/>
	<arg name="mot_pos_en" default="true"/>
	<arg name="light_en" default="true"/>
	<arg name="floor_en" default="true"/>
	<arg name="sel_en" default="true"/>
	<arg name="mot_speed_en" default="true"/>
	<arg name="mic_en" default="true"/>
	<arg name="dist_sens_en" default="true"/>
	<arg name="xpos" default="0.0"/>
	<arg name="ypos" default="0.0"/>
	<arg name="theta" default="0.0"/>
	<arg name="is_single_robot" default="1" />

	<node pkg="epuck_driver" type="epuck2_driver.py" name="$(arg epuck2_name)" output="screen" respawn="true">
		<param name="epuck2_address" value="$(arg epuck2_address)"/>
		<param name="epuck2_name" value="$(arg epuck2_name)"/>
		<param name="camera" value="$(arg cam_en)"/>
		<param name="imu" value="$(arg imu_en)"/>
		<param name="proximity" value="$(arg prox_en)"/>
		<param name="motor_position" value="$(arg mot_pos_en)"/>
		<param name="light" value="$(arg light_en)"/>
		<param name="floor" value="$(arg floor_en)"/>
		<param name="selector" value="$(arg sel_en)"/>
		<param name="motor_speed" value="$(arg mot_speed_en)"/>
		<param name="microphone" value="$(arg mic_en)"/>
		<param name="distance_sensor" value="$(arg dist_sens_en)"/>
		<param name="xpos" value="$(arg xpos)"/>
		<param name="ypos" value="$(arg ypos)"/>
		<param name="theta" value="$(arg theta)"/>
	</node>

	<param name="robot_description" textfile="$(find epuck_driver)/urdf/epuck_urdf.xml"/>

	<node name="epuck_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
		<param name="tf_prefix" value="$(arg epuck2_name)" />
	</node>

	<node if="$(arg is_single_robot)" pkg="rviz" type="rviz" output="screen" name="rviz" args="-d $(find epuck_driver)/config/single_epuck2_driver_rviz.rviz"/>

    <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen" >
        <param name="base_frame" value="$(arg epuck2_name)/base_link"/>
        
        <param name="map_update_interval" value="1.0"/>
        
        <param name="linearUpdate" value="0.01"/> <!-- 1 cm -->
        <param name="angularUpdate" value="0.09"/> <!-- About 5 degrees -->
        <!-- Process a scan if the last scan processed is older than the update time in seconds. A value less than zero will turn time based updates off. -->
        <param name="temporalUpdate" value="-1.0"/>
        
        <!-- map size -->
        <param name="xmin" value="-1.0"/>
        <param name="xmax" value="1.0"/>
        <param name="ymin" value="-1.0"/>
        <param name="ymax" value="1.0"/>
    
        <!-- Processing parameters (resolution of the map)  -->
        <param name="delta" value="0.005"/>
        <param name="maxRange" value="0.080"/>
        <param name="maxUrange" value="0.075"/> 
                        
        <!-- The sigma used by the greedy endpoint matching  -->
        <param name="sigma" value="0.05"/> <!-- default = 0.05 -->
    
        <!-- The kernel in which to look for a correspondence -->
        <param name="kernelSize" value="1"/> <!-- default = 1 -->
                
        <!-- The optimization step in translation -->
        <param name="lstep" value="0.01"/> <!-- default = 0.05 (5 cm) -->
    
        <!-- The optimization step in rotation -->
        <param name="astep" value="0.02"/> <!-- About 1 degree, default = 0.05 (rad) -->    
    
        <!-- The number of iterations of the scanmatcher  -->
        <param name="iterations" value="5"/> <!-- default = 5 -->
    
        <!-- The sigma of a beam used for likelihood computation  -->
        <param name="lsigma" value="0.075"/> <!-- default = 0.075 -->
    
        <!-- Gain to be used while evaluating the likelihood, for smoothing the resampling effects -->
        <param name="ogain" value="3.0"/> <!-- default = 3.0 -->
    
        <!-- Number of beams to skip in each scan. -->
        <param name="lskip" value="0"/> <!-- default = 0 -->
    
        <!-- Minimum score for considering the outcome of the scan matching good. Can avoid jumping -->
        <!-- pose estimates in large open spaces when using laser scanners with limited range (e.g. 5m). -->
        <!-- Scores go up to 600+, try 50 for example when experiencing jumping estimate issues. . -->
        <param name="minimumScore" value="100.0"/> <!-- default = 0.0 -->
    
        <!-- Odometry error in translation as a function of translation (rho/rho) -->
        <param name="srr" value="0.01"/> <!-- default = 0.1 -->
    
        <!-- Odometry error in translation as a function of rotation (rho/theta) -->
        <param name="srt" value="0.02"/> <!-- default = 0.2 -->
    
        <!-- Odometry error in rotation as a function of translation (theta/rho) -->
        <param name="str" value="0.01"/> <!-- default = 0.1 -->
    
        <!-- Odometry error in rotation as a function of rotation (theta/theta) -->
        <param name="stt" value="0.02"/> <!-- default = 0.2 -->
    
        <!-- The Neff based resampling threshold  -->
        <param name="resampleThreshold" value="0.5"/> <!-- default = 0.5 -->
    
        <!-- Number of particles in the filter  -->
        <param name="particles" value="30"/> <!-- default = 30 -->
    
        <!-- Translational sampling range for the likelihood  -->
        <param name="llsamplerange" value="0.01"/> <!-- default = 0.01 -->
    
        <!-- Translational sampling step for the likelihood -->
        <param name="llsamplestep" value="0.01"/> <!-- default = 0.01 -->
    
        <!-- Angular sampling range for the likelihood -->
        <param name="lasamplerange" value="0.005"/> <!-- default = 0.005 -->
    
        <!-- Angular sampling step for the likelihood  -->
        <param name="lasamplestep" value="0.005"/> <!-- default = 0.005 -->
    
        <!-- Threshold on gmapping's occupancy values. Cells with greater occupancy are considered occupied -->
        <!-- (i.e., set to 100 in the resulting sensor_msgs/LaserScan). -->
        <param name="occ_thresh" value="0.25"/> <!-- default = 0.25 -->

        <!-- How long (in seconds) between transform publications.  -->
        <param name="transform_publish_period" value="0.05"/> <!-- default = 0.05 -->
                    
    </node>

</launch>
